File: src/cpp/src/preprocessor/asm/pager.cpp
// SPDX-License-Identifier: Apache-2.0
// Copyright (C) 2024, Advanced Micro Devices, Inc. All rights reserved.

#include "asm/pager.h"
#include "aiebu_error.h"
#include "utils.h"

namespace aiebu {

template <typename T>
std::vector<T>
pager::
union_of_lists_inorder(std::vector<T>& vec1, std::vector<T>& vec2)
{
  // Template to merge unique elements of 2 vector
  std::vector<T> vec3;
  std::merge(vec1.begin(), vec1.end(), vec2.begin(), vec2.end(), std::back_inserter(vec3));
  auto pte = std::unique(vec3.begin(), vec3.end());
  vec3.erase(pte, vec3.end());
  return std::move(vec3);
}

offset_type
pager::
datasectionaligner(offset_type size)
{
  // calculate allignment bytes between text and data section
  return (((size + EOF_SIZE) % DATA_SECTION_ALIGNMENT) > 0) ?
         (DATA_SECTION_ALIGNMENT - ((size + EOF_SIZE) % DATA_SECTION_ALIGNMENT)) : 0;
}

offset_type
pager::
getdatasectionsize(assembler_state& state, std::vector<std::string>& labels_list)
{
  // get data section size for label list
  offset_type dsize16 = 0;
  offset_type dsize4 = 0;
  for (auto &lb : labels_list)
  {
    auto token = state.m_data[state.m_labelmap[lb]->get_index() + 1];
    auto &name = token->get_operation()->get_name();
    if (ALIGNMAP.count(name) > 0)
      if (16 == ALIGNMAP.at(name))
        dsize16 += state.m_labelmap[lb]->get_size();
      else if (4 == ALIGNMAP.at(name))
        dsize16 += state.m_labelmap[lb]->get_size();
      else
        throw error(error::error_code::internal_error, name +" found in ALIGNMAP with allignment " + std::to_string(ALIGNMAP.at(name)));
    else
      throw error(error::error_code::internal_error, lb +" not found in ALIGNMAP!!!");
  }
  return dsize4 + dsize16;
}

std::vector<jobid_type>
pager::
extractjobs(assembler_state& state, std::shared_ptr<job> pjob)
{
  // Extract all jobs depending on jobid
  std::vector<jobid_type> jobids;
  jobids.push_back(pjob->get_jobid());
  std::vector<barrierid_type> barriers = pjob->m_barrierids;
  auto eopnum = pjob->get_eopnum();
  uint32_t i = 0;
  while( i < barriers.size())
  {
    auto lbid = barriers[i];
    auto jj = state.m_localbarriermap[lbid];
    for (auto j : jj)
    {
      if (eopnum != state.m_jobmap[j]->get_eopnum())
        throw error(error::error_code::internal_error, "Job " + std::to_string(j)
                    + " having dependency with Job " + std::to_string(jobids[0]) +
                    " are not expected to be on same page because of .eop");
      barriers = union_of_lists_inorder<barrierid_type>(barriers, state.m_jobmap[j]->m_barrierids);
    }
    jobids = union_of_lists_inorder<jobid_type>(jobids, state.m_localbarriermap[lbid]);
    ++i;
  }
  // Explicit dependencies (e.g. launch_job relation)
  jobids = union_of_lists_inorder<jobid_type>(jobids, pjob->m_dependentjobs);
  return std::move(jobids); 
}

std::vector<std::string>
pager::
extractlabels(assembler_state& state, std::shared_ptr<asm_data> token)
{
  // Extract all labels connected to token
  std::vector<std::string> labels;
  if (token->isLabel())
    return std::move(labels);

  for (auto &arg : token->get_operation()->get_args())
  {
    if (arg.rfind("@") == 0)
    {
      auto lb = arg.substr(1);
      if (state.m_labelmap.find(lb) == state.m_labelmap.end())
        throw error(error::error_code::internal_error, "Label not found " + lb);

      std::vector<std::string> vlb {lb};
      labels = union_of_lists_inorder<std::string>(labels, vlb);
      auto index = state.m_labelmap[lb]->get_index();
      for (auto i = 0; i <= state.m_labelmap[lb]->get_count(); ++i)
      {
        auto elb = extractlabels(state, state.m_data[index+i]);
        labels = union_of_lists_inorder<std::string>(labels, elb);
      }
    }
  }
  return std::move(labels);
}

offset_type
pager::
extractjobsandlabels(assembler_state& state, std::shared_ptr<job> pjob,
                     std::vector<jobid_type>& page_jobs,
                     std::vector<jobid_type>& job_list,
                     std::vector<std::string>& labels_list)
{
  /*
   * extract all releated(jobs which share local barrier with current jobid) jobs
   * and labels with jobid also return size of complete text section in that page
   */

  job_list = extractjobs(state, pjob);
  offset_type tsize = 0;
  for (auto djid : job_list)
  {
    tsize +=  pjob->get_end() - pjob->get_start();
    for (auto j = state.m_jobmap[djid]->get_start_index(); j < state.m_jobmap[djid]->get_end_index() + 1; ++j)
    {
      auto token = state.m_data[j];
      auto elb = extractlabels(state, token);
      labels_list = union_of_lists_inorder<std::string>(labels_list, elb);
    }
  }
  return tsize;
}

std::vector<std::string>
pager::
labelalignmentsorter(assembler_state& state, std::vector<std::string>& clist)
{
  // sort labels in order of there align
  std::vector<std::string> labels;
  for (auto lb : clist)
  {
    auto index = state.m_labelmap[lb]->get_index();
    auto token = state.m_data[index+1];
    if (16 == ALIGNMAP.at(token->get_operation()->get_name()))
      labels.emplace_back(lb);
  }

  for (auto lb : clist)
  {
    auto index = state.m_labelmap[lb]->get_index();
    auto token = state.m_data[index+1];
    if (4 == ALIGNMAP.at(token->get_operation()->get_name()))
      labels.emplace_back(lb);
  }

  if (clist.size() != labels.size())
    throw error(error::error_code::internal_error, "clist size" + std::to_string(clist.size()) + " != " +
          " labels size" + std::to_string(labels.size()));

  return std::move(labels);
}

void
pager::
assignpagenumber(assembler_state& state, uint32_t colnum,
                 std::vector<jobid_type>& jobs,
                 std::vector<std::string> &labels,
                 uint32_t &pagenum, std::vector<page> &pages,
                 bool islastpage)
{
  // assign pagenumber to all instruction and create text and data section lists
  page lpage;
  lpage.set_colnum(colnum);
  lpage.set_pagenum(pagenum);
  lpage.set_islastpage(islastpage);

  if (jobs.size() == 0)
    return;

  if (jobs.size() == 1 && jobs[0] == EOF_ID)
    return;

  // sort labels according to label alignment
  labels = labelalignmentsorter(state, labels);
  for (auto &djid : jobs)
  {
    state.m_jobmap[djid]->set_pagenum(pagenum);
    for (auto j = state.m_jobmap[djid]->get_start_index();
         j < state.m_jobmap[djid]->get_end_index() + 1; ++j)
    {
      state.m_data[j]->set_pagenum(pagenum);
      lpage.m_text.emplace_back(state.m_data[j]);
    }
  }
  for (auto &lb : labels)
  {
    state.m_labelmap[lb]->set_pagenum(pagenum);
    auto index = state.m_labelmap[lb]->get_index();
    for (auto i = 0; i <= state.m_labelmap[lb]->get_count(); ++i)
    {
      state.m_data[index+i]->set_pagenum(pagenum);
      lpage.m_data.emplace_back(state.m_data[index+i]);
    }
  }
  // add eof at end of page
  lpage.m_text.emplace_back(state.m_data[state.m_jobmap[EOF_ID]->get_start_index()]);
  // add align for data section
  lpage.m_text.emplace_back(std::make_shared<asm_data>(std::make_shared<operation>(".align", "16"),
                                                    operation_type::op, code_section::text, 0,
                                                    (uint32_t)-1, 0, ".align 16"));
  pages.emplace_back(std::move(lpage));
  ++pagenum;
}

void
pager::
pagify(assembler_state& state, uint32_t col, std::vector<page>& pages)
{
  // pagify the content
  offset_type pos = PAGE_HEADER_SIZE;
  offset_type page_tsize = 0;
  offset_type page_dsize = 0;
  uint32_t page_index = 0;
  std::vector<jobid_type> page_jobs;
  std::vector<std::string> page_labels;
  auto joblist = state.get_job_list();

  for( auto jobid : state.m_jobids)
  {
    auto pjob = state.m_jobmap[jobid];
    std::vector<jobid_type> job_list;
    std::vector<std::string> labels_list;
    auto it = std::find(page_jobs.begin(), page_jobs.end(), jobid);
    // job already added if job have page assigned or job taken in current page or jodid is 'eof'
    if (pjob->get_pagenum() != NO_PAGE || it != page_jobs.end() || jobid == EOF_ID)
      continue;

    if (jobid <= EOP_ID)
    {
      // conclude the page
      assignpagenumber(state, col, page_jobs, page_labels, page_index, pages, false);
      page_tsize = page_dsize = 0;
      page_jobs.clear();
      page_labels.clear();
      continue;
    }

    // get total text section size, jobs and labels releated to jobid(current job) which are not already taken in current page
    // don't return jobs/labels for a job which is already in this page a second time (note that multiple jobs might depend on
    // the same job)
    auto tsize = extractjobsandlabels(state, pjob, page_jobs, job_list, labels_list);

    // calculate alignment bytes needed bet text and data section
    // NOTE: data section is always 16 Byte aligned
    auto dsectionaligner = datasectionaligner(page_tsize + tsize);

    // get data section size for jobs(related to current job)
    auto dsize = getdatasectionsize(state, labels_list);

    auto jobdsectionaligner = datasectionaligner(tsize);

    //check if job can fit in one page
    if ((tsize + jobdsectionaligner + dsize + EOF_SIZE + PAGE_HEADER_SIZE) > m_page_size)
      throw error(error::error_code::internal_error, "text and data section size " +
                  std::to_string(tsize + jobdsectionaligner + dsize + EOF_SIZE + PAGE_HEADER_SIZE)
                  + " > pagesize(" + std::to_string(m_page_size) + ")");

    // check if current job fit in current page along with other jobs already in page
    if ((page_tsize + page_dsize + tsize + dsectionaligner + dsize + EOF_SIZE + PAGE_HEADER_SIZE) > m_page_size)
    {
      // if current job dont fit in current page, conclude the page
      assignpagenumber(state, col, page_jobs, page_labels, page_index, pages, false);
      page_tsize = page_dsize = 0;
      page_jobs.clear();
      page_labels.clear();
    }

    page_tsize += tsize;
    page_dsize += dsize;
    page_jobs.insert( page_jobs.end(), job_list.begin(), job_list.end() );
    page_labels.insert( page_labels.end(), labels_list.begin(), labels_list.end() );
  }

  if (page_jobs.size())
    assignpagenumber(state, col, page_jobs, page_labels, page_index, pages, true);
  else
    pages[page_index-1].set_islastpage(true);
}

}

File: src/python/aiebu/ctrlcode/pager/pager.py
# SPDX-License-Identifier: MIT
# Copyright (C) 2023-2024 Advanced Micro Devices, Inc.

import math

from ctrlcode.common.section import Section
from ctrlcode.common.page import Page
from ctrlcode.common.util import union_of_lists_inorder
from ctrlcode.common.operation import Operation

class Pager:
    """" pager class """

    HEADER_SIZE = 16
    EOF_SIZE = 4
    DATA_SECTION_ALIGNMENT = 16
    ALIGNMAP = {"uc_dma_bd" : 16, "uc_dma_bd_shim" : 16, ".long": 4}
    OOO = ["load_pdi", "preemption_checkpoint"]

    def __init__(self, pagesize):
        self.pagesize = pagesize
        self.pages = []

    def extractlabels(self, token, state, content):
        """ extract all labels connected to token """
        labels = []
        if not isinstance(token, Operation):
            return labels

        for arg in token.args:
            if arg.startswith('@') and token.name not in Pager.OOO:
                label = arg[1:]   # scratchpad label is not having file scope
                if state.containscratchpads(label):
                    continue
                label = state.getlabelkey(label, token.filename, True)
                assert state.containlabel(label), f"Label not found: {label}"
                labels = union_of_lists_inorder(labels, [label])
                index = state.getlabel(label).getindex()
                # label_index[label]['count'] contain how many lines are there associated with label
                # +1 is for line with label
                for i in range(state.getlabel(label).getcount()+1):
                    labels = union_of_lists_inorder(labels, self.extractlabels(content[index+i].gettoken(), state, content))
        return labels

    def extract_externallabels(self, token, state, content):
        """ extract all external labels connected to token """
        labels = []
        if not isinstance(token, Operation):
            return labels

        for arg in token.args:
            if arg.startswith('@') and token.name in Pager.OOO:
                label = state.getlabelkey(arg[1:], token.filename, True)
                labels = union_of_lists_inorder(labels, [label])
        return labels

    def extractjobs(self, jobid, state):
        """ extract all jobs depending on jobid """
        jobids = [jobid]
        job = state.getjob(jobid)

        # Implicit dependencies by mutual barrier
        barriers = job.getbarrierids().copy()
        blen = int(len(barriers))
        eop_number = job.geteopnumber()
        i = int(0)
        while i < blen:
            lbid = barriers[i]
            for j in state.getlocalbarrierwithid(lbid).getjobids():
                if eop_number != state.getjob(j).geteopnumber():
                    raise RuntimeError(f"Job {j} having dependency with Job {jobids} are not expected"
                                        + " to be on same page because of .eop")
                barriers = union_of_lists_inorder(barriers, state.getjob(j).getbarrierids())
                blen = int(len(barriers))
            jobids = union_of_lists_inorder(jobids, state.getlocalbarrierwithid(lbid).getjobids())
            i += 1

        # Explicit dependencies (e.g. launch_job relation)
        jobids = union_of_lists_inorder(jobids, job.getdependentjobs())

        return jobids

    def assignpagenumber(self, state, col, jobs, labels, external_labels, content, text, data, page_index, pages, islastpage, page_tsize, page_dsize):
        """ assign pagenumber to all instruction and create text and data section lists """
        # sort lables according to label alignment
        labels = self.labelalignmentsorter(labels, state, content)

        for djid in jobs:
            state.getjob(djid).setpagenum(page_index)
            for j in range(state.getjob(djid).getstartindex(),
                           state.getjob(djid).getendindex()+1):
                content[j].setpagenum(page_index)
                text.append(content[j])
        for label in labels:
            state.getlabel(label).setpagenum(page_index)
            index = state.getlabel(label).getindex()
            for i in range(state.getlabel(label).getcount() + 1):
                content[index + i].setpagenum(page_index)
                data.append(content[index + i])

        # add eof at end of last page
        text.append(content[state.getjob('eof').getstartindex()])
        page_rsize = page_tsize + page_dsize + (self.datasectionaligner(page_tsize) if page_dsize else 0)+ Pager.HEADER_SIZE + Pager.EOF_SIZE
        page_rsize = (((page_rsize + 3) >> 2) << 2) # round off to next multiple of 4
        pages.append(Page(col, page_index, text, data, islastpage, page_rsize, 0, external_labels))
        if (len(pages) > 1):
            pages[len(pages) - 2].in_order_page_len = page_rsize

    def extractjobsandlabels(self, state, content, jobid, skip_jobids):
        """ extract all releated(jobs which share local barrier with current jobid) jobs
            and labels with jobid also return size of complete text section in that page """
        job = state.getjob(jobid)
        labels = []
        external_labels = []
        tsize = 0
        jobs = [x for x in self.extractjobs(jobid, state) if x not in skip_jobids]
        for djid in jobs:
            tsize += state.getjob(djid).getend()-state.getjob(djid).getstart()
            for j in range(state.getjob(djid).getstartindex(),
                           state.getjob(djid).getendindex()+1):
                token = content[j].gettoken()
                labels = union_of_lists_inorder(labels, self.extractlabels(token, state, content))
                external_labels = union_of_lists_inorder(external_labels, self.extract_externallabels(token, state, content))
        return tsize, jobs, labels, external_labels

    def labelalignmentsorter(self, clist, state, content):
        """ sort all 16 byte aligned labels together and 4 byte aligned labels together """
        labels = []
        for label in clist:
            index = state.getlabel(label).getindex()
            token = content[index + 1].gettoken()
            if Pager.ALIGNMAP[token.name] == 16:
                labels += [label]

        for label in clist:
            index = state.getlabel(label).getindex()
            token = content[index + 1].gettoken()
            if Pager.ALIGNMAP[token.name] == 4:
                labels += [label]

        assert len(labels) == len(clist), \
            f"len(labels) {len(labels)} != len(clist) {len(clist)}"
        return labels

    def getdatasectionsize(self, labels, state, content):
        """ get data section size for label list """
        dsize16 = 0
        dsize4 = 0
        for label in labels:
            token = content[state.getlabel(label).getindex() + 1].gettoken()
            if token.name in Pager.ALIGNMAP and 16 == Pager.ALIGNMAP[token.name]:
                dsize16 += state.getlabel(label).getsize()
            elif token.name in Pager.ALIGNMAP and 4 == Pager.ALIGNMAP[token.name]:
                dsize4 += state.getlabel(label).getsize()
            else:
                raise RuntimeError(f"Known Data section {token.name}")
        return dsize16 + dsize4

    def datasectionaligner(self, size):
        return (Pager.DATA_SECTION_ALIGNMENT - ((size + Pager.EOF_SIZE) % \
                Pager.DATA_SECTION_ALIGNMENT)) if ((size + Pager.EOF_SIZE) %
                Pager.DATA_SECTION_ALIGNMENT) > 0 else 0

    def pagify(self, state, col, content, relative_page_index=0):
        """ pagify the content """
        state.pos = Pager.HEADER_SIZE
        pagesize = Pager.HEADER_SIZE + Pager.EOF_SIZE
        page_index = relative_page_index
        pages = []
        data = []
        text = []
        page_dsize = 0
        page_tsize = 0
        page_jobs = []
        page_labels = []
        page_external_labels = []
        dsectionaligner = 0
        for jobid in state.getjoblist():
            job = state.getjob(jobid)
             # job already added if job have page assigned or job taken in current page or jodid is 'eof'
            if job.getpagenum() != None or jobid in page_jobs or jobid == 'eof':
                continue

            if type(jobid) == str and jobid[0:3] == 'eop':
                if len(page_jobs):
                    # conclude the page
                    self.assignpagenumber(state, col, page_jobs, page_labels, page_external_labels, content, text, data, page_index, pages, 0, page_tsize, page_dsize)

                    page_index += 1
                    data = []
                    text = []
                    page_tsize = 0
                    page_dsize = 0
                    page_jobs = []
                    page_labels = []
                    page_external_labels = []
                    dsectionaligner = 0
                continue

            # get total text section size, jobs and labels releated to jobid(current job) which are not already taken in current page
            # don't return jobs/labels for a job which is already in this page a second time (note that multiple jobs might depend on
            # the same job)
            tsize, jobs, labels, external_labels = self.extractjobsandlabels(state, content, jobid, page_jobs)

            # calculate alignment bytes needed bet text and data section
            # NOTE: data section is always 16 Byte aligned
            dsectionaligner = self.datasectionaligner(page_tsize + tsize)

            # get data section size for jobs(related to current job)
            dsize = self.getdatasectionsize(labels, state, content)

            jobdsectionaligner = self.datasectionaligner(tsize)

            #print(f"Pagnum:{page_index}\n jobs:{jobs}\n labels:{labels} \nexternal_labels:{external_labels} \ntsize {tsize} + dsize {dsize} + jobdsectionaligner{jobdsectionaligner} + eof{Pager.EOF_SIZE} + page header{Pager.HEADER_SIZE} > Pagesize {self.pagesize}")

            # this assert is to check if job can fit in one page
            assert tsize + jobdsectionaligner + dsize + Pager.EOF_SIZE + Pager.HEADER_SIZE <= self.pagesize, \
                f"tsize {tsize} + dsize {dsize} + jobdsectionaligner{jobdsectionaligner} + eof{Pager.EOF_SIZE} \
                  + page header{Pager.HEADER_SIZE} > Pagesize {self.pagesize}"

            # check if current job fit in current page along with other jobs already in page
            if page_tsize + page_dsize + tsize + dsize + dsectionaligner + Pager.EOF_SIZE + Pager.HEADER_SIZE > self.pagesize:

                # if current job dont fit in current page, conclude the page
                self.assignpagenumber(state, col, page_jobs, page_labels, page_external_labels, content, text, data, page_index, pages, 0, page_tsize, page_dsize)

                page_index += 1
                data = []
                text = []
                page_tsize = 0
                page_dsize = 0
                page_jobs = []
                page_labels = []
                page_external_labels = []
                dsectionaligner = 0

            page_tsize += tsize
            page_dsize += dsize
            page_jobs += jobs
            page_labels += labels
            page_external_labels += external_labels

        if len(page_jobs):
            self.assignpagenumber(state, col, page_jobs, page_labels, page_external_labels, content, text, data, page_index, pages, 1, page_tsize, page_dsize)
            page_index += 1
        else:
            pages[page_index-relative_page_index-1].islastpage = 1
        return page_index, pages
